File: domain/src/main/java/com/movie/domain/repository/ReservationRepository.java
package com.movie.domain.repository;

import com.movie.domain.entity.Reservation;
import com.movie.domain.entity.ReservationStatus;
import com.movie.domain.entity.Schedule;
import com.movie.domain.entity.Seat;
import com.movie.domain.entity.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

public interface ReservationRepository extends JpaRepository<Reservation, Long> {
    @Query("SELECT CASE WHEN COUNT(r) > 0 THEN true ELSE false END FROM Reservation r JOIN r.seats s WHERE r.schedule.id = :scheduleId AND s.id = :seatId")
    boolean existsByScheduleIdAndSeatId(@Param("scheduleId") Long scheduleId, @Param("seatId") Long seatId);

    @Query("SELECT COUNT(r) FROM Reservation r WHERE r.user.id = :userId AND r.schedule.id = :scheduleId AND r.status = :status")
    long countByUserIdAndScheduleIdAndStatus(@Param("userId") Long userId, @Param("scheduleId") Long scheduleId, @Param("status") ReservationStatus status);

    @Query("SELECT r FROM Reservation r WHERE r.user.id = :userId AND r.schedule.id = :scheduleId AND r.status = :status")
    List<Reservation> findByUserIdAndScheduleIdAndStatus(@Param("userId") Long userId, @Param("scheduleId") Long scheduleId, @Param("status") ReservationStatus status);

    @Query("SELECT r FROM Reservation r WHERE r.user.id = :userId")
    List<Reservation> findByUserId(@Param("userId") Long userId);

    Optional<Reservation> findByReservationNumber(String reservationNumber);

    List<Reservation> findBySchedule(Schedule schedule);
} 

File: movie/core/src/main/kotlin/com/example/movie/domain/reservation/repository/ReservationRepository.kt
package com.example.movie.domain.reservation.repository

import com.example.movie.domain.reservation.model.Reservation
import com.example.movie.domain.screening.model.Screening
import com.example.movie.domain.seat.model.Seat

interface ReservationRepository {
    fun findByScreeningAndSeats(screening: Screening, seats: List<Seat>) : List<Reservation>
    fun findByScreeningIdAndUserId(screeningId: Long, userId: Long): List<Reservation>
    fun saveAll(reservations: List<Reservation>): List<Reservation>
    fun findByScreeningAndSeatsWithLock(screening: Screening, seats: List<Seat>): List<Reservation>
}